feat(isIdentityCard): Add PL locale

wiktorwojcik112 opened the pull request:
<!--
Add a descriptive title textbox above, e.g.
feat(validatorName): brief title of what has been done
-->

<!--- briefly describe what you have done in this PR --->

Add isPESEL validation.

PESEL is a polish national identification number. Validator checks whether provided string follows criteria of
valid PESEL.

## Checklist

- [x] PR contains only changes related; no stray files, etc.
- [x] README updated (where applicable)
- [x] Tests written (where applicable)

fedeci reviewed (changes requested):
Does it have something to do with `isIdentityCard`? I am quite in disagreement with adding a specific validator for a specific country code.

wiktorwojcik112 commented:
> Does it have anything to do with `isIdentityCard`? I am quite in disagreement with adding a specific validator for a specific country code.

I checked and it would fit there. I will add this PESEL validator to isIdentityCard.

tux-tn reviewed (changes requested):
Thank you for your PR @wiktorwojcik112 
Totally agree with @fedeci this validator should be added as a locale to `isIdentityCard`. Please make the necessary changes and add tests for [the uncovered line](https://app.codecov.io/gh/validatorjs/validator.js/compare/1745/diff#D2L52) in your code

fedeci commented on src/lib/isIdentityCard.js:
```diff
@@ -1,6 +1,48 @@
+
+      // Ignored because handling else is unnecesary
+      /* istanbul ignore else */
+      if ((modulo === 0 && lastDigit === 0) || (lastDigit === 10 - modulo)) {
```
If you ignore the else the function returns nothing, but it has to return a boolean value

wiktorwojcik112 replied:
Shouldn't it pass through and return false?

wiktorwojcik112 replied:
Ok. I added` return false` outside `else` and I think it should be good now.

tux-tn replied:
Actually the else will be ignored in coverage calculation only, i don't agree with that specially that in your case the missing coverage was in the case of `(lastDigit === 10 - modulo)`

fedeci commented on src/lib/isIdentityCard.js:
```diff
@@ -1,6 +1,48 @@
+        return true;
+      }
+
+      return false;
```
nit: now that the `return false` below is no longer in the else branch we can fallback to it removing this statement
```suggestion
```

wiktorwojcik112 replied:
Shouldn't this stay? If checks at line 22 fail there would be nothing returned.

fedeci replied:
It would fallback to the return on line 44

wiktorwojcik112 replied:
Ok. Thank you for clarification.

fedeci replied:
No problem😄

tux-tn reviewed (changes requested):
Thank you for updating your PR @wiktorwojcik112 can you please fix the linter issue and add the missing test cases to bring back coverage to 100%.
I suggest you run tests locally before pushing new code (using `yarn test` or `npm run test`)

wiktorwojcik112 commented:
Ok. I fixed linter error and changed title, but I still can't figure out how to solve coverage issue.

tux-tn commented:
> Ok. I fixed linter error and changed title, but I still can't figure out how to solve coverage issue.

That's tricky but you need to find a test example where the last digit of PESEL number equals 10 minus your modulo. 
As i told you earlier the uncovered execution path is `lastDigit === 10 - modulo`

wiktorwojcik112 commented:
> > Ok. I fixed linter error and changed title, but I still can't figure out how to solve coverage issue.
> 
> That's tricky but you need to find a test example where the last digit of PESEL number equals 10 minus your modulo. As i told you earlier the uncovered execution path is `lastDigit === 10 - modulo`

Everything should be good now.

tux-tn commented on src/lib/isIdentityCard.js:
```diff
@@ -1,6 +1,46 @@
+    };
+
+    if (str != null && str.length === 11 && isInt(str, { allow_leading_zeroes: true })) {
+      const digits = str.split('');
+      let sum = 0;
+
+      digits.forEach((digit, index) => {
+      // Ignored because handling else is unnecesary
+      /* istanbul ignore else */
+        if (index !== 10) {
+          sum += Number(digit) * weightOfDigits[index + 1];
+        }
+      });
```
I'm pretty sure your `istanbul ignore else` is unneeded here. BTW i suggest using `Array.reduce` instead of forEach here
```suggestion
      const digits = str.split('').slice(0, -1);
      const sum = digits.reduce((acc, digit, index) =>
        acc + (Number(digit) * weightOfDigits[index + 1]), 0);
```

wiktorwojcik112 replied:
I agree, so I have removed `Istanbul ignore else` and replaced Array.forEach with Array.reduce.

tux-tn reviewed (changes requested):
Thank you @wiktorwojcik112 can you please address my latest comment concerning `istanbul ignore else` ?

tux-tn reviewed (approved):
LGTM 🎉  Thank you for taking the time to quickly address all the comments

wiktorwojcik112 commented:
> LGTM 🎉 Thank you for taking the time to quickly address all the comments

It was nice working with you. Thank you for help. Take care!

profnandaa reviewed (approved):
LGTM. thanks for your contrib! 🎉
